sound/soc/sof/ops.h

```diff
@@ -77,7 +77,9 @@ static inline int snd_sof_dsp_core_power_up(struct snd_sof_dev *sdev,
 static inline int snd_sof_dsp_core_power_down(struct snd_sof_dev *sdev,
 					      unsigned int core_mask)
 {
-	if (sof_ops(sdev)->core_power_down)
+	/* Only power down the enabled cores */
+	core_mask = core_mask & sdev->enabled_cores_mask;
```

Comment:
maybe a paranoid remark but is there a need to protect sdev->enabled_cores_mask? Could there be races where different cores are powered-up and down simultaneously?

Reply:
@bardliao this likely wont help the issue because the enabled_cores_mask is only update when powering up or powering down the cores during topology loading/unloading. I see 2 problems: 1. we dont use the snd_sof_dsp_power_up/down wrapper everywhere for this to take effect. For example: in hda-dsp_remove(), we use hda_dsp_core_reset_power_down() directly. 2. enabled_cores_mask is not updated during firmware boot.

Reply:
> maybe a paranoid remark but is there a need to protect sdev->enabled_cores_mask? Could there be races where different cores are powered-up and down simultaneously?

@plbossart it's an interesting question. So far I don't see any locking used to protect that mask (or other involved fields and objects). E.g. for PM when calling `sof_resume()` -> `sof_restore_pipelines()` -> `sof_load_pipeline_ipc()` -> `sof_core_enable()`. So, just adding a lock here wouldn't help. Incidentally I only see `snd_sof_dsp_core_power_down()` being called on error paths so far, so, don't think this patch addresses an issue that we really observe?
And BTW, you could use `&=` ;-)

Reply:
> @bardliao this likely wont help the issue because the enabled_cores_mask is only update when powering up or powering down the cores during topology loading/unloading. I see 2 problems: 1. we dont use the snd_sof_dsp_power_up/down wrapper everywhere for this to take effect. For example: in hda-dsp_remove(), we use hda_dsp_core_reset_power_down() directly. 2. enabled_cores_mask is not updated during firmware boot.

enabled_cores_mask is also update in snd_sof_run_firmware(). So I think problem 2 is not a problem. For problem 1, I have no idea why we use hda_dsp_core_reset_power_down() directly instead of snd_sof_dsp_core_power_down(). Should we replace all the hda_dsp_core_reset_power_down() by snd_sof_dsp_core_power_down()? 
The other concern to me is that we didn't update ``sdev->enabled_cores_mask`` in snd_sof_dsp_core_power_up() and snd_sof_dsp_core_power_down(). Is this a bug or on purpose?

Reply:
> And BTW, you could use `&=` ;-)

Thanks, I will update it.

Reply:
> The other concern to me is that we didn't update `sdev->enabled_cores_mask` in snd_sof_dsp_core_power_up() and snd_sof_dsp_core_power_down(). Is this a bug or on purpose?

multi-core support has not really been tested extensively and AFAIK we've never enabled it on any topology yet. But yes, we should fix it. And replace hda_dsp_core_reset_power_down/up with the top-level op and make these static so they dont get abused.